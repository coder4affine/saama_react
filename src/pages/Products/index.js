import React, { useEffect } from 'react';

import { connect } from 'react-redux';
import PropTypes from 'prop-types';
// import { makeStyles } from '@material-ui/core/styles';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableContainer from '@material-ui/core/TableContainer';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';
import Paper from '@material-ui/core/Paper';
import Button from '@material-ui/core/Button';
import IconButton from '@material-ui/core/IconButton';
import EditIcon from '@material-ui/icons/Edit';
import DeleteIcon from '@material-ui/icons/Delete';
// import { ProductContext } from '../../contex/ProductsContext';

export const Index = ({
  history,
  loadProducts,
  deleteProductRequest,
  products: { loading, data: products, error },
}) => {
  // const { products, loading, error, loadProducts } = useContext(ProductContext);
  useEffect(() => {
    if (products.length <= 0) {
      loadProducts();
    }
  }, []);

  if (loading) {
    return <h3>Loading...</h3>;
  }

  if (error) {
    return <h3>{error.message}</h3>;
  }

  return (
    <div>
      <h1>Products Page</h1>
      <Button
        data-testid="btn-add-product"
        variant="contained"
        color="primary"
        onClick={() => history.push('addProduct')}
      >
        Add Product
      </Button>

      <TableContainer component={Paper}>
        <Table aria-label="simple table">
          <TableHead>
            <TableRow>
              <TableCell>Product Name</TableCell>
              <TableCell align="right">Product Price</TableCell>
              <TableCell align="right">Manufacturer</TableCell>
              <TableCell>Actions</TableCell>
            </TableRow>
          </TableHead>
          <TableBody>
            {products.map((x) => (
              <TableRow key={x.id}>
                <TableCell component="th" scope="row">
                  {x.productName}
                </TableCell>
                <TableCell align="right">{x.productPrice}</TableCell>
                <TableCell align="right">{x.manufacturer}</TableCell>
                <TableCell>
                  <div>
                    <IconButton
                      data-testid={`btn-edit-product-${x.id}`}
                      aria-label="edit"
                      onClick={() => {
                        history.push(`/updateProduct/${x.id}`);
                      }}
                    >
                      <EditIcon />
                    </IconButton>
                    <IconButton aria-label="delete" onClick={() => deleteProductRequest(x.id)}>
                      <DeleteIcon />
                    </IconButton>
                  </div>
                </TableCell>
              </TableRow>
            ))}
          </TableBody>
        </Table>
      </TableContainer>
    </div>
  );
};

Index.propTypes = {
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
  }).isRequired,
  loadProducts: PropTypes.func.isRequired,
  products: PropTypes.shape({
    loading: PropTypes.bool.isRequired,
    data: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        productName: PropTypes.string.isRequired,
        productPrice: PropTypes.number.isRequired,
        manufacturer: PropTypes.string.isRequired,
      }).isRequired,
    ).isRequired,
    error: PropTypes.instanceOf(Error),
  }).isRequired,
  deleteProductRequest: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => ({
  products: state.products,
});

const mapDispatchToProps = (dispatch) => ({
  loadProducts: () => dispatch({ type: 'LOAD_PRODUCTS_REQUEST' }),
  deleteProductRequest: (payload) => dispatch({ type: 'DELETE_PRODUCT_REQUEST', payload }),
});

export default connect(mapStateToProps, mapDispatchToProps)(Index);
