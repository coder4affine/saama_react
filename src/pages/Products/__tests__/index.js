import 'isomorphic-fetch';
import React from 'react';
import { render, waitFor, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import { Router } from 'react-router-dom';
import customHistory from '../../../history';
import configureStore from '../../../configureStore';
import Products from '../index';

const mockPush = jest.fn();

const mockHistory = {
  push: mockPush,
};

const renderWithRedux = (component, { store = configureStore } = {}) => ({
  ...render(
    <Provider store={store}>
      <Router history={customHistory}>{component}</Router>
    </Provider>,
  ),
  store,
});

describe('Products Component', () => {
  let mockAPI;
  beforeEach(() => {
    const mockSuccessResponse = [
      {
        id: 1,
        productName: 'iphone',
        productPrice: 70000,
        manufacturer: 'Apple',
      },
    ];
    const mockJsonPromise = Promise.resolve(mockSuccessResponse);
    const mockFetchPromise = Promise.resolve({
      json: () => mockJsonPromise,
    });

    mockAPI = jest.spyOn(global, 'fetch').mockImplementationOnce(() => mockFetchPromise);
  });

  afterEach(() => {
    mockPush.mockClear();
  });

  test('should render products component', () => {
    const { container } = renderWithRedux(<Products history={mockHistory} />, {
      store: configureStore,
    });

    expect(container).toMatchSnapshot();
  });

  test('should render products after api success', async () => {
    const { container } = renderWithRedux(<Products history={mockHistory} />, {
      store: configureStore,
    });

    await waitFor(() => expect(mockAPI).toHaveBeenCalledTimes(1));

    expect(container).toMatchSnapshot();
  });

  test('should click add product button', async () => {
    const { getByTestId } = renderWithRedux(<Products history={mockHistory} />, {
      store: configureStore,
    });

    await waitFor(() => expect(mockAPI).toHaveBeenCalledTimes(1));

    fireEvent.click(getByTestId('btn-add-product'));
    expect(mockPush).toHaveBeenCalledTimes(1);
    expect(mockPush).toHaveBeenCalledWith('addProduct');
  });

  test('should click edit product button', async () => {
    const { getByTestId } = renderWithRedux(<Products history={mockHistory} />, {
      store: configureStore,
    });

    await waitFor(() => expect(mockAPI).toHaveBeenCalledTimes(1));

    fireEvent.click(getByTestId('btn-edit-product-1'));
    expect(mockPush).toHaveBeenCalledTimes(1);
    expect(mockPush).toHaveBeenCalledWith('/updateProduct/1');
  });
});
