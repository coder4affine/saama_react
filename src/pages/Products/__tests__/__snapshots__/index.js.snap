// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Products Component should render products after api success 1`] = `
<div>
  <div>
    <h1>
      Products Page
    </h1>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
      data-testid="btn-add-product"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Add Product
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <div
      class="MuiPaper-root MuiTableContainer-root MuiPaper-elevation1 MuiPaper-rounded"
    >
      <table
        aria-label="simple table"
        class="MuiTable-root"
      >
        <thead
          class="MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head"
              scope="col"
            >
              Product Name
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight"
              scope="col"
            >
              Product Price
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight"
              scope="col"
            >
              Manufacturer
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head"
              scope="col"
            >
              Actions
            </th>
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root"
        >
          <tr
            class="MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-body"
              scope="row"
            >
              iphone
            </th>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
            >
              70000
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
            >
              Apple
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body"
            >
              <div>
                <button
                  aria-label="edit"
                  class="MuiButtonBase-root MuiIconButton-root"
                  data-testid="btn-edit-product-1"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
                <button
                  aria-label="delete"
                  class="MuiButtonBase-root MuiIconButton-root"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`Products Component should render products component 1`] = `
<div>
  <h3>
    Loading...
  </h3>
</div>
`;
